File: android/src/org/tmurakam/cashflow/models/AssetEntry.java
// -*-  Mode:java; c-basic-offset:4; tab-width:4; indent-tabs-mode:t -*-

package org.tmurakam.cashflow.models;

import java.lang.*;
import java.util.*;

import android.database.*;
import android.database.sqlite.*;

import org.tmurakam.cashflow.ormapper.*;
import org.tmurakam.cashflow.models.*;

//
// 各資産（勘定）のエントリ
//

class AssetEntry implements Cloneable {
    public int assetKey;
    public double value;
    public double balance;

    public Transaction transaction;

    public AssetEntry() {
        transaction = null;
        assetKey = -1;
        value = 0.0;
        balance = 0.0;
    }

    public AssetEntry(Transaction t, Asset asset) {
        assetKey = asset.pid;
        if (t == null) {
            // 新規エントリ生成
            transaction = new Transaction();
            transaction.asset = assetKey;
        } else {
            transaction = t;

            if (isDstAsset()) {
                value = -t.value;
            } else {
                value = t.value;
            }
        }
    }

    //
    // 資産間移動の移動先取引なら YES を返す
    //
    public boolean isDstAsset() {
        if (transaction.type == Transaction.TRANSFER && assetKey == transaction.dst_asset) {
            return true;
        }
        return false;
    }

    public Transaction getTransaction() {
        _setupTransaction();
        return transaction;
    }


    // 値を Transaction に書き戻す
    private void _setupTransaction() {
        if (transaction.type == Transaction.ADJ) {
            transaction.balance = balance;
            transaction.hasBalance = true;
        } else {
            transaction.hasBalance = false;
            if (isDstAsset()) {
                transaction.value = -value;
            } else {
                transaction.value = value;
            }
        }
    }

    // TransactionViewController 用の値を返す
    public double evalue() {
        double ret = 0.0;

        switch (transaction.type) {
        case Transaction.INCOME:
            ret = value;
            break;
        case Transaction.OUTGO:
            ret = -value;
            break;
        case Transaction.ADJ:
            ret = balance;
            break;
        case Transaction.TRANSFER:
            if (isDstAsset()) {
                ret = value;
            } else {
                ret = -value;
            }
            break;
        }
	
        if (ret == 0.0) {
            ret = 0.0;	// avoid '-0'
        }
        return ret;
    }

    // 編集値をセット
    public void setEvalue(double v) {
        switch (transaction.type) {
        case Transaction.INCOME:
            value = v;
            break;
        case Transaction.OUTGO:
            value = -v;
            break;
        case Transaction.ADJ:
            balance = v;
            break;
        case Transaction.TRANSFER:
            if (isDstAsset()) {
                value = v;
            } else {
                value = -v;
            }
            break;
        }
    }

    // 種別変更
    //   type のほか、transaction の dst_asset, asset, value も調整する
    public boolean changeType(int type, int assetKey, int dstAssetKey) {
        if (type == Transaction.TRANSFER) {
            if (dstAssetKey == this.assetKey) {
                // 自分あて転送は許可しない
                // ### TBD
                return false;
            }

            transaction.type = Transaction.TRANSFER;
            setDstAsset(dstAssetKey);
        } else {
            // 資産間移動でない取引に変更した場合、強制的に指定資産の取引に変更する
            double ev = evalue();
            transaction.type = type;
            transaction.asset = assetKey;
            transaction.dst_asset = -1;
            setEvalue(ev);
        }
        return true;
    }

    // 転送先資産のキーを返す
    public int dstAsset() {
        if (transaction.type != Transaction.TRANSFER) {
            //ASSERT(false);
            return -1;
        }

        if (isDstAsset()) {
            return transaction.asset;
        }

        return transaction.dst_asset;
    }

    public void setDstAsset(int assetKey) {
        if (transaction.type != Transaction.TRANSFER) {
            //ASSERT(false);
            return;
        }

        if (isDstAsset()) {
            transaction.asset = assetKey;
        } else {
            transaction.dst_asset = assetKey;
        }
    }

    public Object clone() {
        AssetEntry e = new AssetEntry();
        e.assetKey = assetKey;
        e.value = value;
        e.balance = balance;
        e.transaction = (Transaction)transaction.clone();
        return e;
    }
}

File: Classes/models/AssetEntry.m
// -*-  Mode:ObjC; c-basic-offset:4; tab-width:8; indent-tabs-mode:nil -*-
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2011, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

// AssetEntry

#import "AppDelegate.h"
#import "Asset.h"

@implementation AssetEntry

@synthesize assetKey = mAssetKey;
@synthesize transaction = mTransaction;
@synthesize value = mValue;
@synthesize balance = mBalance;

- (void)dealloc
{
    [mTransaction release];
    [super dealloc];
}

- (id)init
{
    self = [super init];

    mTransaction = nil;
    mAssetKey = -1;
    mValue = 0.0;
    mBalance = 0.0;

    return self;
}

- (id)initWithTransaction:(Transaction *)t withAsset:(Asset *)asset
{
    self = [self init];

    mAssetKey = asset.pid;
    
    if (t == nil) {
        // 新規エントリ生成
        self.transaction = [[[Transaction alloc] init] autorelease];
        mTransaction.asset = self.assetKey;
    }
    else {
        self.transaction = t;

        if ([self isDstAsset]) {
            mValue = -t.value;
        } else {
            mValue = t.value;
        }
    }

    return self;
}

//
// 資産間移動の移動先取引なら YES を返す
//
- (BOOL)isDstAsset
{
    if (mTransaction.type == TYPE_TRANSFER && mAssetKey == mTransaction.dstAsset) {
        return YES;
    }

    return NO;
}

- (Transaction *)transaction
{
    [self _setupTransaction];
    return mTransaction;
}

// 値を Transaction に書き戻す
- (void)_setupTransaction
{
    if (mTransaction.type == TYPE_ADJ) {
        mTransaction.balance = mBalance;
        mTransaction.hasBalance = YES;
    } else {
        mTransaction.hasBalance = NO;
        if ([self isDstAsset]) {
            mTransaction.value = -mValue;
        } else {
            mTransaction.value = mValue;
        }
    }
}

// TransactionViewController 用の値を返す
- (double)evalue
{
    double ret = 0.0;

    switch (mTransaction.type) {
    case TYPE_INCOME:
        ret = mValue;
        break;
    case TYPE_OUTGO:
        ret = -mValue;
        break;
    case TYPE_ADJ:
        ret = mBalance;
        break;
    case TYPE_TRANSFER:
        if ([self isDstAsset]) {
            ret = mValue;
        } else {
            ret = -mValue;
        }
        break;
    }
	
    if (ret == 0.0) {
        ret = 0.0;	// avoid '-0'
    }
    return ret;
}

// 編集値をセット
- (void)setEvalue:(double)v
{
    switch (mTransaction.type) {
    case TYPE_INCOME:
        mValue = v;
        break;
    case TYPE_OUTGO:
        mValue = -v;
        break;
    case TYPE_ADJ:
        mBalance = v;
        break;
    case TYPE_TRANSFER:
        if ([self isDstAsset]) {
            mValue = v;
        } else {
            mValue = -v;
        }
        break;
    }
}

// 種別変更
//   type のほか、transaction の dst_asset, asset, value も調整する
- (BOOL)changeType:(int)type assetKey:(int)as dstAssetKey:(int)das
{
    if (type == TYPE_TRANSFER) {
        if (das == self.assetKey) {
            // 自分あて転送は許可しない
            // ### TBD
            return NO;
        }

        mTransaction.type = TYPE_TRANSFER;
        [self setDstAsset:das];
    } else {
        // 資産間移動でない取引に変更した場合、強制的に指定資産の取引に変更する
        double ev = self.evalue;
        mTransaction.type = type;
        mTransaction.asset = as;
        mTransaction.dstAsset = -1;
        self.evalue = ev;
    }
    return YES;
}

// 転送先資産のキーを返す
- (int)dstAsset
{
    if (mTransaction.type != TYPE_TRANSFER) {
        ASSERT(NO);
        return -1;
    }

    if ([self isDstAsset]) {
        return mTransaction.asset;
    }

    return mTransaction.dstAsset;
}

- (void)setDstAsset:(int)as
{
    if (mTransaction.type != TYPE_TRANSFER) {
        ASSERT(NO);
        return;
    }

    if ([self isDstAsset]) {
        mTransaction.asset = as;
    } else {
        mTransaction.dstAsset = as;
    }
}

- (id)copyWithZone:(NSZone *)zone
{
    AssetEntry *e = [[AssetEntry alloc] init];
    e.assetKey = self.assetKey;
    e.value = self.value;
    e.balance = self.balance;
    e.transaction = [[self.transaction copy] autorelease];

    return e;
}

@end
